Add database tests to explain performance.

JoshDunnDev opened the pull request.

JoshDunnDev commented on lib/batchVersions.js:
```diff
@@ -249,12 +273,18 @@ export async function insertOptions({options}) {
+
   const records = await collection.find(
-    {}, {projection, sort}).toArray();
+    {}, {projection, sort}).limit(1).toArray();
```
`limit(1)` is added here to ensure that we only examine 1 index key. If `limit(1)` is not added, it will examine each `batchVersion.id` key for every record in the database.

dlongley replied:
Yeah, we should have had that `limit(1)` there from the start.

dlongley commented on lib/batchVersions.js:
```diff
@@ -139,7 +139,7 @@ export async function ensureBatchVersion({tokenizerId}) {
  *
  * @returns {object} A tokenizerVersion object.
```
This returns something different when `explain` is passed -- so that should be documented.

JoshDunnDev replied:
I went ahead an updated this. Not sure if this is exactly what you were looking for but let me know.

dlongley commented on lib/tokens.js:
```diff
@@ -1015,6 +1003,39 @@ async function _getOpenBatch({
 }
 
+export async function _updateTokenBatch({
+  batchId,
+  compressed,
+  resolvedList,
+  encodedRequester,
+  requesterList,
+  explain = false
```
We probably don't need to single line this as it just takes up more space.

dlongley commented on lib/tokens.js:
```diff
@@ -1059,9 +1079,7 @@ async function _getUnfilledBatch({
-  const query = {'tokenBatch.id': batchId};
-  const projection = {_id: 0};
-  let record = await collection.findOne(query, {projection});
+  let record = await _getBatch({id: batchId});
```
This breaks the code because it will throw if the `batchId` is not found. Previously, this did not call `_getBatch` (which throws on not found) to avoid that. So this will need to be fixed in some way to ensure it behaves the same way it used to.

dlongley commented on lib/documents.js:
```diff
@@ -29,6 +29,10 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
     fields: {'registration.externalIdHash': 1, 'registration.documentHash': 1},
     options: {unique: true, background: false}
+  }, {
+    collection: 'tokenization-registration',
+    fields: {'registration.internalId': 1},
+    options: {unique: true, background: false}
```
So this is tricky.

We previously determined that we couldn't add this index for performance reasons. Sharding is on `externalIdHash` and this does not include it and this would cause scatter-gather queries (and our query cannot include `externalIdHash` in `getRegistration`). This shouldn't be a problem on its own (I don't think), but, IIRC, this index creates insertion performance problems. We should re-evaluate this unless someone else better remembers the details.

I thought this was better documented in the `getRegistration` call to indicate that it will always be an unindexed slow call -- but apparently not. Regardless of whether we add this index, we should add documentation to that function to indicate that it is a low performance (potentially very low) method -- either it will do a full doc scan or it will do scatter-gather (based on whether we can add this index).

If we end up adding this index we should also add a comment here indicating that it doesn't include the shard key so it will result in scatter-gather queries (and we know that and it's ok). If we opt to have no index here we should also indicate why we don't have one.

dlongley commented on lib/documents.js:
```diff
@@ -26,6 +26,8 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
     // and queries must be supported to find them all...
     // but do not allow multiple `internalId`s for the same `hash` and
     // `externalIdHash`
+    // an index is not created for registration.internalId due to performance
```
```suggestion
    // Note: an index is not created for `registration.internalId` because queries
    // based on that field alone are rare; they are not run in hot code paths nor do
    // they need to run quickly -- so the index cost is not justified
```

dlongley commented on lib/documents.js:
```diff
@@ -26,6 +26,8 @@ bedrock.events.on('bedrock-mongodb.ready', async () => {
     // but do not allow multiple `internalId`s for the same `hash` and
     // `externalIdHash`
+    // an index is not created for registration.internalId due to performance
+    // issues related to sharding.
```
```suggestion
```

dlongley commented on lib/documents.js:
```diff
@@ -290,12 +292,21 @@ export async function _generateInternalId() {
  */
-export async function getRegistration({internalId} = {}) {
+export async function getRegistration({internalId, explain = false} = {}) {
+  // this query will be slow due to not having an index on
+  // registration.internalId.
```
```suggestion
  // this query will be slow due to *intentionally* not having an index on
  // `registration.internalId`; see index creation comments above
```

dlongley commented on lib/documents.js:
```diff
@@ -290,12 +292,21 @@ export async function _generateInternalId() {
  *
```
Please add a sentence to the docs for this function above after `Retrieves the encrypted registration...`. It should say "This function may be extremely slow to execute, so do not call it in any hot code paths."

dlongley commented on test/mocha/helpers.js:
```diff
@@ -6,6 +6,7 @@
 const database = require('bedrock-mongodb');
 
 exports.isRegistration = result => {
+  // FIXME: add proper result checking instead of a console log.
```
Will this be addressed in this PR? If not, it should have a link to an issue.

JoshDunnDev replied:
I did create an issue for this since it's unrelated to this PR: https://github.com/digitalbazaar/bedrock-tokenization/issues/69

dlongley commented on test/mocha/mock.data.js:
```diff
@@ -24,3 +23,99 @@ for(const product of products) {
+    tokenizerId: 'did:key:z6MkugLpCsat2iV8Xb3Y9kwUf3bMFFexXr4jUbs75grMkofo',
+    jwe: {
+      protected: 'eyJlbmMiOiJYQzIwUCJ9',
+      recipients: [Array],
```
This looks invalid.

dlongley reviewed (changes requested):
I did not see a test for `getRegistration` that confirms there's no index being used -- did I just overlook it?

JoshDunnDev commented:
> I did not see a test for `getRegistration` that confirms there's no index being used -- did I just overlook it?

Yes I just updated the old test I had for it and added a comment to the test as well.

dlongley commented on test/mocha/10-documents.js:
```diff
@@ -30,240 +36,308 @@ const key2 = new X25519KeyAgreementKey2020({
+    });
+    it(`is NOT indexed for 'registration.internalId' in getRegistration()`,
+      async function() {
+      // Note: an index is not created for `registration.internalId` because
+      // queries based on that field alone are rare; they are not run in hot
+      // code paths nor do they need to run quickly -- so the index cost is not
+      // justified. Therefore, a collection scan should be done.
```
```suggestion
        // Note: an index is not created for `registration.internalId` because
        // queries based on that field alone are rare; they are not run in hot
        // code paths nor do they need to run quickly -- so the index cost is not
        // justified. Therefore, a collection scan should be done.
```

dlongley commented on test/mocha/10-documents.js:
```diff
@@ -30,240 +36,308 @@ const key2 = new X25519KeyAgreementKey2020({
+      executionStats.executionStages.inputStage.inputStage.inputStage.stage
+        .should.equal('IXSCAN');
+    });
+    it(`is properly indexed for compound query of ` +
+    `'registration.externalIdHash' and 'registration.documentHash' in ` +
+    `_refresh()`, async function() {
```
```suggestion
    it('is properly indexed for compound query of ' +
      `'registration.externalIdHash' and 'registration.documentHash' in ` +
      '_refresh()', async function() {
```

dlongley commented on test/mocha/10-documents.js:
```diff
@@ -30,240 +36,308 @@ const key2 = new X25519KeyAgreementKey2020({
-      result2 = await documents._hmacString({
-        hmac,
-        value: '0e26c923-84e6-4918-9337-f82c56951007'
+    it(`is properly indexed for compound query of ` +
+    `'registration.externalIdHash' and 'registration.documentHash' in ` +
+    `_getRegistrationRecord()`, async function() {
```
Here and elsewhere needs some indentation and quote fixing:

```suggestion
    it('is properly indexed for compound query of ' +
      `'registration.externalIdHash' and 'registration.documentHash' in ` +
      '_getRegistrationRecord()', async function() {
```

dlongley reviewed (approved):
Thanks! LGTM other than some style nits in the tests. I didn't see that other comment or the test before because github hid it due to the diff size (so my search didn't work).
